text/0000-panic-in-drop.md

```diff
@@ -0,0 +1,332 @@
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Panics which happen inside a `Drop` impl currently unwind as long as the drop was not itself called during unwinding (which would otherwise abort immediated due to a double-panic). This RFC proposes to change this behavior to always abort if a panic attempts to escape a drop, even if this drop was invoked as part of normal function execution (as opposed to unwinding which already aborts with a double-panic).
+
+# Motivation
+[motivation]: #motivation
+
+## Exception safety
+
+Unsafe code must be [exception-safe](https://doc.rust-lang.org/nomicon/exception-safety.html). That is, it must ensure that it does not expose any potential unsoundness to safe code even when an unwinding panic occurs in the middle of it. This is particularly tricky to handle when the unsafe call must call user-provided functions that may panic, usually via trait methods. This is *exceptionally* tricky in the case of panics caused from dropping user-provided types since drops are usually invisible and implicit in the source code.
+
+Forgetting to take panics from drop into account has been the source of several bugs in popular crates and even the standard library:
+- Double drop in the standard library: https://github.com/rust-lang/rust/issues/83618
+- Double drop in `arrayvec`: https://github.com/bluss/arrayvec/issues/3
+- The exact same double-drop bug in `smallvec`: https://github.com/servo/rust-smallvec/issues/14
+
+### `catch_unwind`
+
+One particularly nasty case of missed exception safety is [rust-lang/rust#86027](https://github.com/rust-lang/rust/issues/86027).
+
+Code using `catch_unwind` is not typically prepared to handle an object that panics in its `Drop` impl. Even the standard library has had serious bugs in this regard, and if the standard library doesn't consistently get it right, we can hardly expect others to do so.
+
+`catch_unwind(code)` is often used to make sure no panics from `code` can cause further unwinding/panics. However, when catching a panic with a payload that panics on Drop, most usages of `catch_unwind(code)` will still result in further unwinding and often unsoundness.
+
+Example in the standard library (issue [rust-lang/rust#86030](https://github.com/rust-lang/rust/issues/86030)):
+
+```rust
+struct Bomb;
+
+impl Drop for Bomb {
+    fn drop(&mut self) {
+        panic!();
+    }
+}
+
+std::panic::panic_any(Bomb);
+```
+
+https://github.com/rust-lang/rust/blob/5ea19239d9d6f49fdd76513a36386d7e83708e3f/library/std/src/rt.rs#L34-L39
+
+```rust
+fn main() {
+    std::panic::panic_any(Bomb);
+}
+```
+```
+thread 'main' panicked at 'Box<Any>', src/main.rs:12:5
+note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
+thread 'main' panicked at 'explicit panic', src/main.rs:7:9
+fatal runtime error: failed to initiate panic, error 5
+abort (core dumped)
+```
```

Comment:
By the way, this whole section is not really relevant anymore since we can have https://github.com/rust-lang/rust/pull/99032